cupyx/scipy/signal/_iir_utils.py

```diff
@@ -0,0 +1,220 @@
+        return;
+    }
+
+    if(carry_stream != NULL) {
+        cudaStreamWaitEvent(carry_stream, carry_event, 0);
+    }
+
+    const double* prev_carries = carries + (n_group - 1) * k;
+    double* this_group = out + group_start;
+
+    double carry = 0.0;
+    for(int i = 1; i <= k; i++) {
+        const double* k_factors = factors + (m + k) * (i - 1) + k;
+        double factor = k_factors[idx];
+        double k_value = prev_carries[k - i];
+        carry += factor * k_value;
+    }
+
+    this_group[idx] += carry;
+
+}
+
+"""
+
+IIR_MODULE = cp.RawModule(
+    code=IIR_KERNEL, options=('-std=c++11', '-dc'),
+    name_expressions=['compute_correction_factors',
+                      'first_pass_iir',
+                      'copy_carries',
+                      'second_pass_iir'])
+
+
+def apply_iir(x, b):
+    x = x.astype(cp.float64)
+    b = b.astype(cp.float64)
+    out = x.copy()
+
+    k = b.size
+    n = x.size
+    block_sz = 32
+    n_blocks = (n + block_sz - 1) // block_sz
+
+    correction = cp.eye(k)
+    correction = cp.c_[correction[::-1], cp.empty((k, block_sz))]
+    carries = cp.empty((n_blocks - 1, k), dtype=cp.float64)
+
+    corr_kernel = IIR_MODULE.get_function('compute_correction_factors')
+    first_pass_kernel = IIR_MODULE.get_function('first_pass_iir')
+    copy_kernel = IIR_MODULE.get_function('copy_carries')
+    second_pass_kernel = IIR_MODULE.get_function('second_pass_iir')
+
+    corr_kernel((k,), (1,), (block_sz, k, b, correction))
+    first_pass_kernel((n_blocks,), (block_sz // 2,),
+                      (block_sz, k, n, correction, out, carries))
+
+    streams = [None]
+    streams += [cp.cuda.Stream(non_blocking=True) for _ in range(n_blocks - 1)]
+    events = [cp.cuda.Event() for _ in streams]
+
+    exec_streams = [cp.cuda.Stream() for _ in range(n_blocks - 2)]
+
+    default_stream = cp.cuda.get_current_stream()
+
+    exec_streams += [default_stream]
+    exec_events = [cp.cuda.Event(block=True) for _ in exec_streams]
+
+    for i in range(n_blocks - 1):
+        exec_stream = exec_streams[i]
+        exec_event = exec_events[i]
+        cur_stream = streams[i + 1]
+        prev_stream = streams[i]
+        cur_event = events[i + 1]
+        prev_event = events[i]
+        prev_stream_ptr = prev_stream.ptr if prev_stream is not None else 0
+        prev_event_ptr = prev_event.ptr if prev_event is not None else 0
+        with cur_stream:
+            copy_kernel((1,), (k,),
+                        (block_sz, k, n, i + 1, correction, carries,
+                         out, prev_stream_ptr, prev_event_ptr))
+            cur_event.record(cur_stream)
+
+        with exec_stream:
+            second_pass_kernel((1,), (block_sz,),
+                               (block_sz, k, n, i + 1, correction, carries,
+                                out, prev_stream_ptr, prev_event_ptr))
+            exec_event.record(exec_stream)
+
+    for evt in exec_events:
+        default_stream.wait_event(evt)
```

Comment:
@takagi, I was wondering if you know if the events and streams are being synchronized here effectively? I'm having sync issues, as it is evident by the tests; sometimes some combination fails and then others, but not consistently. When `CUDA_LAUNCH_BLOCKING` is set to 1, then the tests always pass.

### Random run where all tests pass
![imagen](https://user-images.githubusercontent.com/1878982/222824632-df75686b-a9cb-4f86-9bfb-805360a3b4b0.png)

### Inconsistent failures
![imagen](https://user-images.githubusercontent.com/1878982/222825029-453a6a4c-51c6-4acb-8325-8e8ddd6024a3.png)

![imagen](https://user-images.githubusercontent.com/1878982/222825094-9fb7987a-f7f1-4d60-ad1d-ba115ea71ebf.png)


### Blocking streams
![imagen](https://user-images.githubusercontent.com/1878982/222824529-09378e04-abc5-40c6-b298-f4534c4052d1.png)

Reply:
@andfoy  I have big concerns regarding this. You're mixing the stream ordering semantics on host & device, which nothing guarantees that it'd work. In particular, the [Programming Guide says](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#events-cdp1)
> As for all device runtime objects, event objects may be shared between all threads within the thread-block which created them but are local to that block and may not be passed to other kernels, or between blocks within the same kernel. Event handles are not guaranteed to be unique between blocks, so using an event handle within a block that did not create it will result in undefined behavior.

That is, if you do dynamic parallelism, the events should be created on device, not on host and be used on device. What you're doing now is simply UB. Any reason you don't just do `cudaStreamWaitEvent` on host?

Reply:
Also doing `-dc`/`-rdc` (relocatable code) is a bit costly, and unless dynamic parallelism is absolutely needed in (i.e. an integral part of) your algorithm, you should avoid it.

Reply:
I'll try to convert the current execution pipeline into dynamic parallelism